### ISSUE TITLE
Unreliable caching for multistage dockerfiles when stages are based on previous stages

### ISSUE THREAD

**Issue by twuttk** [GoogleContainerTools/kaniko#575]
**Actual behavior**
Caching for multistage dockerfiles with stages based on previous stages is unreliable and often doesn't work properly. Minimal working example below.

Caching fails at stage 2:
```
FROM debian:stable-slim AS stage1
RUN apt-get update
RUN touch stageone_file

FROM stage1 As stage2
RUN apt-get update
RUN touch stagetwo_file
```

Caching works:
```
FROM debian:stable-slim AS stage1
#RUN apt-get update
RUN touch stageone_file

FROM stage1 As stage2
RUN apt-get update
RUN touch stagetwo_file
```

Caching works:
```
FROM debian:stable-slim AS stage1
RUN apt-get update
RUN touch stageone_file

FROM stage1 As stage2
#RUN apt-get update
RUN touch stagetwo_file
```

Please note that this is a different issue than described and fixed in https://github.com/GoogleContainerTools/kaniko/issues/458

**Expected behavior**
Caching should always function properly, regardless of the amount of stages and their relation to each other.

**To Reproduce**
Build images with dockerfiles quoted above

**Additional Information**
- this was tested with the latest debug image.

**Comment by dlorenc** [comment]
Hey,

Thanks for the report! I'm able to reproduce this with the images tagged :latest, but not with any built from HEAD. Could you try using one of these images and see if it still repros?

gcr.io/kaniko-project/executor:4b7e2b3a064fe0d558d2a87173bc50e7de31f441
gcr.io/kaniko-project/executor:debug-4b7e2b3a064fe0d558d2a87173bc50e7de31f441

I'm not sure off the top of my head what recent change would have fixed this, though.

**Comment by temiwottku** [comment]
Hi, 

thanks for the quick answer. For me, using the images built from HEAD doesn't seem to change anything, caching for the example from above still fails at stage 2, while the other to examples work fine.

I'm running kaniko within minikube and mount my code into the container as separate volume. Here is my pod.yaml:

```
apiVer
sion: v1
kind: Pod
metadata:
  name: kaniko
spec:
  containers:
    - name: kaniko
      image: gcr.io/kaniko-project/executor:4b7e2b3a064fe0d558d2a87173bc50e7de31f441
      args: ["--context=/repo",
             "--dockerfile=multistage-caching-fails.dockerfile",
             "--destination=<private registry>",
             "--cache=true"]
      volumeMounts:
        - name: jenkins-docker-cfg
          mountPath: /root
        - name: repo
          mountPath: /repo
  restartPolicy: Never
  volumes:
    - name: jenkins-docker-cfg
      projected:
        sources:
          - secret:
              name: secret
              items:
                - key: .dockerconfigjson
                  path: .docker/config.json
    - name: repo
      hostPath:
        path: /hosthome/<path-to-my-repo>
```


And again the dockerfile which doesn't work to have it all in one place:
```

FROM debian:stable-slim AS stage1
RUN apt-get update
RUN touch stageone_file

FROM stage1 As stage2
RUN apt-get update
RUN touch stagetwo_file
```

**Comment by apenney** [comment]
I opened #682 earlier and I realized it's exactly the same problem as this.  I use a -slim container, do an apt-get as my first action, and then have something fail to cache in stage2.

I can confirm this is still an issue with the latest kaniko container.

**Comment by burdiyan** [comment]
Same thing here. Seems to only cache the first stage in our situation.

**Comment by cvgw** [comment]
This should be fixed in v0.15.0
Please re-open if this problem persists
